src/app/clusters/actions-server/actions-server.cpp

```diff
@@ -0,0 +1,470 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "actions-server.h"
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app-common/zap-generated/ids/Attributes.h>
+#include <app/AttributeAccessInterfaceRegistry.h>
+#include <app/CommandHandlerInterfaceRegistry.h>
+#include <app/EventLogging.h>
+#include <app/util/attribute-storage.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::Actions;
+using namespace chip::app::Clusters::Actions::Attributes;
+using namespace chip::Protocols::InteractionModel;
+
+namespace {
+static constexpr size_t kActionsDelegateTableSize =
+    MATTER_DM_ACTIONS_CLUSTER_SERVER_ENDPOINT_COUNT + CHIP_DEVICE_CONFIG_DYNAMIC_ENDPOINT_COUNT;
+static_assert(kActionsDelegateTableSize <= kEmberInvalidEndpointIndex, "Actions Delegate table size error");
+
+Delegate * gDelegateTable[kActionsDelegateTableSize] = { nullptr };
+
+Delegate * GetDelegate(EndpointId aEndpoint)
+{
+    return (aEndpoint >= kActionsDelegateTableSize ? nullptr : gDelegateTable[aEndpoint]);
+}
```

Comment:
This is a bad pattern to continue. We need to have a better state keeping pattern than constantly using global static arrays.

Reply:
@andy31415 ^^^

Reply:
The right solutions in general seem to be:

1. One cluster instance per endpoint as opposed to having a singleton that handles all endpoint, or
2. Some way that the cluster instance maps endpoints to delegates that can be configured sanely at compile time if we want to support size caps on it, or that just does dynamic memory (e.g. std::map) if that's OK for the cluster in question.

I suspect in most cases item 1 is a better idea....

Reply:
Even with item 1, we would need to map endpoints with the delegates in the server implementations?

Reply:
With item 1 each cluster instance is a separate object which just has a pointer to its delegate; there is nothing to "map".

Reply:
Let's do this in follow-up. Added todo in the code comment.